swupd/rename_test.go

```diff
@@ -0,0 +1,156 @@
+		k string // stuff to make content unique
+	}{
+		// Two files, same contents
+		{n: "I1", f: File{Name: "/lib/python3.6/libpython.3.6.0.so"}, s: 400, m: _rwx, k: "file1"},
+		{n: "I2", f: File{Name: "/lib/python3.7/libpython.3.7.0.so"}, s: 400, m: _rwx, k: "file1"},
+		// Two short files, different contents
+		{n: "S1", f: File{Name: "/lib/python3.6/tiny"}, s: 180, m: _rwx, k: "123"},
+		{n: "S2", f: File{Name: "/lib/python3.7/tiny"}, s: 180, m: _rwx, k: "456"},
+		// long files, different contents, same length
+		{n: "L1", f: File{Name: "/lib/python3.6/big"}, s: 580, m: _rwx, k: "123"},
+		{n: "L2", f: File{Name: "/lib/python3.7/big"}, s: 580, m: _rwx, k: "456"},
+		{n: "L3", f: File{Name: "/lib/python2.6/big"}, s: 580, m: _rwx, k: "123"},
+		{n: "L4", f: File{Name: "/lib/python2.7/big"}, s: 580, m: _rwx, k: "456"},
+	}
+
+	m := &Manifest{}
+	m.Header.Version = 20
+
+	sn := make(map[string]File)
+	for i := range files {
+		f := &files[i]
+		hi := HashFileInfo{Mode: f.m, Size: f.s}
+		contents := make([]byte, f.s)           // Load of NUL of correct length
+		copy(contents[0:len(f.k)], []byte(f.k)) // set the first n bytes
+		hv, err := GetHashForBytes(&hi, contents)
+		if err != nil {
+			t.Fatalf("GetHashForBytes for %v returned %v", f, err)
+		}
+		f.f.Hash = internHash(hv)
+		f.f.Info = sizer(f.s) // A FileInfo that returns our desired size
+		sn[f.n] = f.f
+	}
+	return sn
+}
+
+// filelist takes a list of wanted shortnames, looks them up in sn, and returns
+// the a list of pointers to copies of the Files. Because they are copies they can
+// be altered without side effects on other tests
+func filelist(t *testing.T, sn map[string]File, wanted []string) []*File {
+	r := make([]*File, len(wanted))
+	for i := range wanted {
+		element, ok := sn[wanted[i]]
+		if !ok {
+			t.Fatalf("Shortname %v not found in array", wanted[i])
+		}
+		r[i] = &element
+	}
+	return r
+}
+
+// markdelete sets all the files in the slice to deleted
+func markdelete(a []*File) []*File {
+	for i := range a {
+		a[i].Status = StatusDeleted
+	}
+	return a
+}
+
+// checkLinked tests forward and backward pairing of DeltaPeers
+func checkLinked(t *testing.T, a, b *File, description string) bool {
+	if a.DeltaPeer != b {
+		t.Errorf("%s Incorrect Linkage forward %v -> %v, got %v", description, a, b, a.DeltaPeer)
+		return false
+	}
+	if b != nil && b.DeltaPeer != a {
+		t.Errorf("%s Incorrect Linkage reverse %v -> %v, got %v", description, b, a, b.DeltaPeer)
+		return false
+	}
+	return true
+}
+
+func TestRename(t *testing.T) {
+	sn := genTestArray(t)
+	// for f := range sn {
+	// 	fmt.Printf("%s %v\n", f, sn[f])
+	// }
+	tests := []struct {
+		description string
+		remove      []string // shortnames of File to remove
+		add         []string // shortnames of File to add
+		partner     []int    // Indexes of add that should match remove
+	}{
+		{description: "Rename of file, same contents",
+			remove: []string{"I1", "I2"}, add: []string{"I2"}, partner: []int{0, -1}},
```

Comment:
I1 and I2 are removed but I2 is also added? When could this be possible?....... Oh, shortnames. Also forgot that `sn` is being generated.

I'm leaving this so you can see my thought process, no action needed.